### DISCUSSION TITLE
How to "Ready to Review" / undraft pull request

### DISCUSSION THREAD

**Discussion by kanimaru** [hub4j/github-api#1578] [Q&A] [answered]
Hi a short question how would you undraft a pull request?

Tried with GHPullRequest.requestTeamReviewers but didn't work
Also tried with GHPullRequest.reopen() also not working.

**Comment by gsmet** [comment, marked as answer]
Unfortunately, there's nothing you can do with the REST API, GitHub doesn't expose an endpoint for that (at least to my knowledge, I did another check and couldn't find something about it).

So what you need to do is use the GraphQL API which exposes a mutation for it: https://docs.github.com/en/graphql/reference/mutations#markpullrequestreadyforreview .

Some operations are only available in the REST API and some others are only available in the GraphQL API. That's a bit unfortunate but that's how it is.

I think I remember you asking questions in Quarkus GitHub App so if you are still using it, something like this should probably work (not tested):

```
Map<String, Object> variables = new HashMap<>();
variables.put("pullRequestId", pullRequest.getNodeId());

Response response = gitHubGraphQLClient.executeSync("""
        mutation MarkPullRequestReadyForReview($pullRequestId: ID!) {
          markPullRequestReadyForReview(input: {
            pullRequestId: $pullRequestId }) {
                clientMutationId
          }
        }""", variables);
```

You have a full example here of how to use GraphQL in Quarkus GitHub App given Discussions are only exposed as GraphQL: https://github.com/quarkusio/quarkus-github-bot/blob/main/src/main/java/io/quarkus/bot/TriageDiscussion.java .

HTH.

**Reply by kanimaru** [reply]
I'll mark this as answer. Cause I think for PAT this should be the way to go. For Github Application's it's probably not possible.

**Reply by gsmet** [reply]
I will have a closer look probably next week. Sometimes there are bugs on the GitHub side and they are not checking the right permission.

**Reply by PedroDSFerreira** [reply]
The GitHub App can perform this operation, using the installation ID as the bearer token.

Query:
```
mutation MarkPullRequestReadyForReview {
    markPullRequestReadyForReview(input: { pullRequestId: "<YOUR_PR_ID>" }) {
        clientMutationId
    }
}
```
(tested in Postman)

**Reply by bitwiseman** [reply]
That is the GraphQL API, which this library doesn't support yet.

**Reply by PedroDSFerreira** [reply]
Yes I know, I was saying that this operation works with GH Apps as well, because @kanimaru was unsure if it was possible.

**Comment by kanimaru** [comment]
Almost work like that, 
But now GitHub answers with:

```graphql
GraphQLResponse{
data={"markPullRequestReadyForReview":null}, 
errors=[
  Error{
    message=Resource not accessible by integration, 
    locations=[{line=2, column=3}], 
    path=[markPullRequestReadyForReview], 
    extensions={saml_failure=false}, 
    otherFields={type=FORBIDDEN}}
]
}
```

Code in Use:

```java
    graphql.executeAsync("""
        mutation MarkPullRequestReadyForReview($pullRequestId: ID!) {
          markPullRequestReadyForReview(input: {
            pullRequestId: $pullRequestId
          }) {
            clientMutationId
          }
        }
        """, Map.of("pullRequestId", pullRequest.getNodeId()))
        .subscribe().with(
            response -> log.info("Undraft successfully: {}", response),
            throwable -> log.error("Can't undraft PR", throwable)
        );
```

The App has PR Read/Write, Issue Read/Write, Contents Read/Write access.
Any Idea whats missing?

**Reply by gsmet** [reply]
I would expect PR Read/Write to be enough.

You use the GraphQL client automatically injected, right? Also common mistake: have you approved the permission changes if you have changed the permissions? You usually get an email to approve them if you change them after the creation.

**Reply by gsmet** [reply]
Also the pull request is in a project where the GitHub App is installed?

**Reply by kanimaru** [reply]
> I would expect PR Read/Write to be enough.

Yeah strange, maybe some special microsoft logic / bug

No there is no default bean to inject:
```java
@Inject
DynamicGraphQLClient graphql;
```
didn't work with: `Caused by: java.lang.RuntimeException: URL not configured for client. Please define the property quarkus.smallrye-graphql-client.default.url or pass it to your client builder dynamically`

Used:

```java
  @Produces
  @ApplicationScoped
  public DynamicGraphQLClient graphql(GitHubService gitHubService, GithubConfig config) {
    return gitHubService.getInstallationGraphQLClient(config.installationId());
  }
``` 
to produce my own client.

And yes permissions are all approved.

And yes the Bot is in the same Organization and for this Repository enabled.

**Reply by gsmet** [reply]
Are you doing that when listening to an event?

Because what you should do in this case is just add a `DynamicGraphQLClient graphqlClient` parameter to your method and it will be automatically injected and properly authenticated. But I suppose you would have a correct client the way you did it too even if not recommended.
All I can think of is maybe the installation id being incorrect.

Maybe try my way and see if it works better. See https://github.com/quarkusio/quarkus-github-bot/blob/main/src/main/java/io/quarkus/bot/TriageDiscussion.java#L41-L43 for an example.

**Reply by kanimaru** [reply]
The problem with the approach to get the client at the begin I would need to bubble it completly through the application through all layers until the end when the change will be executed. I tried you version also but same result. The installation id works for REST completely with creating PR answering etc. Just Undrafting doesn't work =,=....

But fine I give up and do it with labels instead. I think this Draft thing was just an: "Oh Gitlab has it. We need it to, please hack it in" thing...
